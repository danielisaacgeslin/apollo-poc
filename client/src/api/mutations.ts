import { gql } from 'apollo-boost';

export const CREATE_CHANNEL = gql`
  mutation($name: String!, $type: String!) {
    createChannel(channel: { name: $name, type: $type }) {
      id
    }
  }
`;

export const DELETE_CHANNEL = gql`
  mutation($id: ID!) {
    deleteChannel(id: $id)
  }
`;
