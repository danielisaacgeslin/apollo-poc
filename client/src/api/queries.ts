import { gql } from 'apollo-boost';

export const GET_FULL_CHANNEL_LIST = gql`
  {
    channels {
      id
      name
      type
    }
  }
`;
