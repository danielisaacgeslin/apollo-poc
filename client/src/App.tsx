import React, { useCallback, useEffect } from 'react';
import { useQuery, useMutation } from '@apollo/react-hooks';

import { Channel } from '../../server/src/channel/ChannelAPI'; /** @todo move to a shared location */
import { GET_FULL_CHANNEL_LIST } from './api/queries';
import { CREATE_CHANNEL, DELETE_CHANNEL } from './api/mutations';
import { mainCointainer } from './styles';
import { ChannelForm } from './ChannelForm';
import { ChannelList } from './ChannelList';

export default () => {
  const { loading, error, data, refetch } = useQuery<{ channels: Channel[] }>(GET_FULL_CHANNEL_LIST);
  const [createChannel, creationStatus] = useMutation<{ createChannel: { id: string } }>(CREATE_CHANNEL);
  const [deleteChannel, deletionStatus] = useMutation<{ deleteChannel: boolean }>(DELETE_CHANNEL);

  const onSave = useCallback((name: string, type: string) => createChannel({ variables: { name, type } }), [createChannel]);
  const onDelete = useCallback((id: string) => deleteChannel({ variables: { id } }), [deleteChannel]);

  useEffect(() => {
    if (!creationStatus.data?.createChannel?.id && !deletionStatus.data?.deleteChannel) return;
    refetch({});
  }, [creationStatus, deletionStatus, refetch]);

  return (
    <div className={mainCointainer}>
      <ChannelForm isLoading={creationStatus.loading} onSave={onSave} />
      <ChannelList isLoading={loading} isDeleting={deletionStatus.loading} channelList={data?.channels as Channel[]} error={error} onDelete={onDelete} />
    </div>
  );
};
