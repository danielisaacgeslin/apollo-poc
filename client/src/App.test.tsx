import React from 'react';
import { MockedProvider, MockedResponse, wait } from '@apollo/react-testing';
import { render, act, fireEvent } from '@testing-library/react';

import { GET_FULL_CHANNEL_LIST } from './api/queries';
import { CREATE_CHANNEL, DELETE_CHANNEL } from './api/mutations';

import { default as App } from './App';

const getWrappedApp = (mocks: MockedResponse[]) => (
  <MockedProvider mocks={mocks} addTypename={false}>
    <App />
  </MockedProvider>
);

describe('App', () => {
  let mocks: MockedResponse[];
  let fetchMock: jest.Mock;
  let createMock: jest.Mock;
  let deleteMock: jest.Mock;
  beforeEach(() => {
    fetchMock = jest.fn();
    createMock = jest.fn();
    deleteMock = jest.fn();
    mocks = [
      {
        request: { query: GET_FULL_CHANNEL_LIST },
        result: () => {
          fetchMock();
          return {
            data: {
              channels: [
                { id: 'a', name: 'first', type: 'dm' },
                { id: 'b', name: 'second', type: 'group' }
              ]
            }
          };
        }
      },
      {
        request: { query: CREATE_CHANNEL, variables: { name: 'new channel', type: 'dm' } },
        result: () => {
          createMock();
          return { data: { createChannel: { id: 'c' } } };
        }
      },
      {
        request: { query: DELETE_CHANNEL, variables: { id: 'a' } },
        result: () => {
          deleteMock();
          return { data: { deleteChannel: true } };
        }
      }
    ];
  });

  it('should render loading', () => {
    const { container } = render(getWrappedApp(mocks));
    expect(container).toMatchSnapshot();
  });

  it('should render with results', async () => {
    const { container } = render(getWrappedApp(mocks));
    await act(() => wait(0));
    expect(container).toMatchSnapshot();
  });

  it('should create a channel', async () => {
    const { getByTestId } = render(getWrappedApp([...mocks, mocks[0]]));
    fireEvent.change(getByTestId('channel-name'), { target: { value: 'new channel' } });
    fireEvent.click(getByTestId('channel-save'));
    await act(() => wait(0));
    expect(createMock).toBeCalled();
  });

  it('should delete a channel', async () => {
    window.confirm = () => true;
    const { getAllByTestId } = render(getWrappedApp([...mocks, mocks[0]]));
    await act(() => wait(0));
    fireEvent.click(getAllByTestId('channel-delete')[0]);
    await act(() => wait(0));
    expect(deleteMock).toBeCalled();
  });
});
