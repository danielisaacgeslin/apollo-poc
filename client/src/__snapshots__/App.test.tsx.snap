// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render loading 1`] = `
.emotion-8 {
  padding: 15px;
}

.emotion-7 {
  padding: 15px 0;
  max-width: 300px;
}

.emotion-6 {
  margin-top: 15px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  margin-bottom: 10px;
}

.emotion-1 {
  height: 30px;
  padding: 2px 5px;
  background-color: #FFFFFF;
  color: #252A2E;
}

.emotion-2 {
  height: 30px;
  background-color: #FFFFFF;
  color: #252A2E;
}

.emotion-4 {
  font-size: 13px;
  padding: 3px 10px;
  background-color: #D6DDF6;
  color: #252A2E;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  background-color: #3AC6AB;
  color: #FFFFFF;
}

.emotion-5 {
  font-size: 13px;
  padding: 3px 10px;
  background-color: #D6DDF6;
  color: #252A2E;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
}

<div>
  <div
    class="emotion-8"
  >
    <div
      class="emotion-7"
    >
      <h4>
        create a new channel
      </h4>
      <div
        class="emotion-3"
      >
        <input
          class="emotion-0 emotion-1"
          data-testid="channel-name"
          placeholder="type a channel name"
          value=""
        />
        <select
          class="emotion-2"
          data-testid="channel-type"
        >
          <option
            value="dm"
          >
            dm
          </option>
          <option
            value="group"
          >
            group
          </option>
        </select>
      </div>
      <div
        class="emotion-6"
      >
        <button
          class="emotion-4"
          data-testid="channel-save"
          disabled=""
          style="margin-right: 5px;"
        >
          save
        </button>
        <button
          class="emotion-5"
          data-testid="channel-cancel"
        >
          cancel
        </button>
      </div>
    </div>
    <div>
      <h4>
        Channel list
      </h4>
      <div>
        <p>
          loading list...
        </p>
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`App should render with results 1`] = `
.emotion-12 {
  padding: 15px;
}

.emotion-7 {
  padding: 15px 0;
  max-width: 300px;
}

.emotion-6 {
  margin-top: 15px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  margin-bottom: 10px;
}

.emotion-9 {
  margin-bottom: 5px;
}

.emotion-1 {
  height: 30px;
  padding: 2px 5px;
  background-color: #FFFFFF;
  color: #252A2E;
}

.emotion-2 {
  height: 30px;
  background-color: #FFFFFF;
  color: #252A2E;
}

.emotion-4 {
  font-size: 13px;
  padding: 3px 10px;
  background-color: #D6DDF6;
  color: #252A2E;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  background-color: #3AC6AB;
  color: #FFFFFF;
}

.emotion-5 {
  font-size: 13px;
  padding: 3px 10px;
  background-color: #D6DDF6;
  color: #252A2E;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
}

.emotion-8 {
  font-size: 13px;
  padding: 3px 10px;
  background-color: #D6DDF6;
  color: #252A2E;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  background-color: #B73F65;
  color: #FFFFFF;
}

<div>
  <div
    class="emotion-12"
  >
    <div
      class="emotion-7"
    >
      <h4>
        create a new channel
      </h4>
      <div
        class="emotion-3"
      >
        <input
          class="emotion-0 emotion-1"
          data-testid="channel-name"
          placeholder="type a channel name"
          value=""
        />
        <select
          class="emotion-2"
          data-testid="channel-type"
        >
          <option
            value="dm"
          >
            dm
          </option>
          <option
            value="group"
          >
            group
          </option>
        </select>
      </div>
      <div
        class="emotion-6"
      >
        <button
          class="emotion-4"
          data-testid="channel-save"
          disabled=""
          style="margin-right: 5px;"
        >
          save
        </button>
        <button
          class="emotion-5"
          data-testid="channel-cancel"
        >
          cancel
        </button>
      </div>
    </div>
    <div>
      <h4>
        Channel list
      </h4>
      <div>
        <div>
          <p
            class="emotion-9"
          >
            <button
              class="emotion-8"
              data-testid="channel-delete"
            >
              x
            </button>
             
            first
             
            <small>
              (
              dm
              )
            </small>
          </p>
          <p
            class="emotion-9"
          >
            <button
              class="emotion-8"
              data-testid="channel-delete"
            >
              x
            </button>
             
            second
             
            <small>
              (
              group
              )
            </small>
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
